tools/cli/commands/clean.js

```diff
@@ -0,0 +1,607 @@
+		fs.rm( file, { recursive: true, force: true }, ( err ) => {
+			console.log( 'Cleaning ', file );
+			if ( err ) {
+				// File deletion failed
+				console.error( err.message );
+				return;
+			}
+		} );
+	}
+
+	// Cleanup root and tools node_modules folder if that's what we're deleting.
+	if ( argv.scope === 'all' && argv.toClean.includes( 'node_modules' ) ) {
+		process.on( 'exit', async () => {
+			await runCommand( 'rm', [ '-rf', 'node_modules', 'tools/cli/node_modules' ] );
+		} );
+	}
+}
+/**
+ * Returns list of files that we want to delete.
+ *
+ * @param {Array} allFiles - a list of all possible deletable files.
+ * @param {Array} toClean - what kind of files we want to delete.
+ * @returns {Array} deleteQueue - files that we want to delete.
+ */
+async function collectCleanFiles( allFiles, toClean ) {
+	const deleteQueue = [];
+	for ( const file of toClean ) {
+		switch ( file ) {
+			case 'untracked':
+				deleteQueue.push( ...allFiles.untracked );
+				break;
+			case 'docker':
+				deleteQueue.push( ...allFiles.docker );
+				break;
+			case 'node_modules':
+				deleteQueue.push( ...allFiles.node_modules );
+				break;
+			case 'composer.lock':
+				deleteQueue.push( ...allFiles.composerLock );
+				break;
+			case 'vendor':
+				deleteQueue.push( ...allFiles.vendor );
+				break;
+			case 'other':
+				deleteQueue.push( ...allFiles.other );
+				break;
+		}
+	}
+
+	for ( const file of deleteQueue ) {
+		console.log( file );
+	}
+
+	return deleteQueue;
+}
+/**
+ * Gets list of files that could be deleted.
+ *
+ * @param {Array} toClean - files that we want to clean.
+ * @returns {object} allFiles.
+ */
+async function collectAllFiles( toClean ) {
+	const allFiles = {
+		untracked: [],
+		docker: [],
+		node_modules: [],
+		vendor: [],
+		composerLock: [],
+		other: [],
+		combined: [],
+	};
+
+	if ( toClean.includes( 'untracked' ) ) {
+		allFiles.untracked = child_process.execSync( 'git ls-files --exclude-standard --directory --other' );
+		allFiles.untracked = allFiles.untracked.toString().trim().split( '\n' );
+	}
+
+	allFiles.other = child_process.execSync( 'git ls-files --exclude-standard --directory --ignored --other' );
+	allFiles.other = allFiles.other.toString().trim().split( '\n' );
+
+	allFiles.combined = allFiles.untracked.concat( allFiles.other );
+
+	for ( const file of allFiles.combined ) {
+		if ( file.match( /^\.env$|^tools\/docker\// ) ) {
+			allFiles.docker.push( file );
+			allFiles.other.splice( allFiles.other.indexOf( file ) );
+		}
+		if ( file.match( /(^|\/)node_modules\/$/ ) ) {
+			allFiles.node_modules.push( file );
+			allFiles.other.splice( allFiles.other.indexOf( file ) );
+		}
+		if ( file.match( /(^|\/)vendor\/$/ ) ) {
+			allFiles.vendor.push( file );
+			allFiles.other.splice( allFiles.other.indexOf( file ) );
+		}
+		if ( file.match( /(^|\/)composer\.lock$/ ) ) {
+			allFiles.composerLock.push( file );
+			allFiles.other.splice( allFiles.other.indexOf( file ) );
+		}
+	}
```

Comment:
This is going to break if we ever wind up with `tools/docker/node_modules/`, `tools/docker/compose.lock`, or the like. First line 185 will remove it from the list, then one of the other lines will wind up doing `allFiles.other.splice( -1 )` which will delete some other random file from `allFiles.other`.

I'd probably do something like this.
```suggestion
	allFiles.combined = child_process.execSync( 'git ls-files --exclude-standard --directory --ignored --other' );
	allFiles.combined = allFiles.combined.toString().trim().split( '\n' );

	for ( const file of allFiles.combined ) {
		if ( file.match( /^\.env$|^tools\/docker\// ) ) {
			allFiles.docker.push( file );
		} else if ( file.match( /(^|\/)node_modules\/$/ ) ) {
			allFiles.node_modules.push( file );
		} else if ( file.match( /(^|\/)vendor\/$/ ) ) {
			allFiles.vendor.push( file );
		} else if ( file.match( /(^|\/)composer\.lock$/ ) ) {
			allFiles.composerLock.push( file );
		} else {
			allFiles.other.push( file );
		}
	}
```

Reply:
38ce25b